### PR TITLE
Improve model load priority

### LINKED ISSUE #834: Add load priority property to model entities
Right now loading order is a bit unpredictable

### PR DESCRIPTION
**Pull request by HifiExperiments**
Closes #834

This contains multiple changes to how we prioritize loading models:
- Avatar model entities have a higher loading priority
- Non-avatar model entities have a dynamically updated priority that changes as you move around and gives priority to things that are in view
- Model entities have a new property, `loadPriority`, that you can set to override the automatically computed priority.  This is a bit experimental, so I have not exposed it to Create.
- In an FST, you can specify `waitForWearables=true` (e.g. https://gist.github.com/HifiExperiments/fbb063e083129fc04530a9e1dd833443/raw/08ee9c9c3d047c23f5e0e6765b84a953c37a120e/waitForWearablesTest.fst), which will make your avatar not show up until your avatar entity wearables have loaded.

This is all a bit difficult to test, because many factors impact loading performance.  For example, the loading priority is only used if we have *pending* loading requests.  So when you first enter a domain, the entity mixer will send you a bunch of entities, and, because we're under the concurrent download limit, some will begin loading right away without checking their priority.  Only once we reach the limit are the priorities used to determine subsequent downloads.  Also, just because a model starts loading first doesn't mean it will finish first.  I'm open to suggestions if there are other recommendations for how to improve this.

- [ ] file follow up issue to support specifying avatar entities in FSTs, and specifying that only some are necessary to wait for

## Funding

This project is funded through [NGI0 Entrust](https://nlnet.nl/entrust), a fund established by [NLnet](https://nlnet.nl) with financial support from the European Commission's [Next Generation Internet](https://ngi.eu) program. Learn more at the [NLnet project page](https://nlnet.nl/project/Overte).

[NLnet foundation logo](https://nlnet.nl)
[NGI Zero Logo](https://nlnet.nl/entrust)

### CODE REVIEW

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -5733,15 +5731,30 @@ void Application::init() {
 
     getEntities()->init();
     getEntities()->setEntityLoadingPriorityFunction([this](const EntityItem& item) {
-        auto dims = item.getScaledDimensions();
-        auto maxSize = glm::compMax(dims);
+        if (item.getEntityHostType() == entity::HostType::AVATAR) {
+            return item.isMyAvatarEntity() ? Avatar::MYAVATAR_ENTITY_LOADING_PRIORITY : Avatar::OTHERAVATAR_ENTITY_LOADING_PRIORITY;
```
Comment (HifiExperiments) [inline interface/src/Application.cpp:5735]: this is where avatar entities are assigned a higher priority (highest for MyAvatar entities, slightly lower for OtherAvatar entities)

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -5733,15 +5731,30 @@ void Application::init() {
...
+        float result = atan2(maxSize, distance);
+
+        bool isInView = true;
+        {
+            QMutexLocker viewLocker(&_viewMutex);
+            isInView = _viewFrustum.sphereIntersectsKeyhole(itemPosition, maxSize);
+        }
+        if (!isInView) {
+            const float OUT_OF_VIEW_PENALTY = -M_PI_2;
+            result += OUT_OF_VIEW_PENALTY;
```
Comment (HifiExperiments) [inline interface/src/Application.cpp:5754]: this is where out of view entities are penalized

File: interface/src/avatar/OtherAvatar.h [hunk 3]
```diff
@@ -91,6 +93,14 @@ class OtherAvatar : public Avatar {
     uint8_t _workloadRegion { workload::Region::INVALID };
     BodyLOD _bodyLOD { BodyLOD::Sphere };
     bool _needsDetailedRebuild { false };
+
+private:
+    // When determining _hasCheckedForAvatarEntities for OtherAvatars, we can set it to true in
+    // handleChangedAvatarEntityData if we have avatar entities.  But we never receive explicit
+    // confirmation from the avatar mixer if we don't have any.  So instead, we wait to receive
+    // a few identity packets, and assume that if we haven't gotten any avatar entities by then,
+    // that we're safe to say there aren't any.
```
Comment (HifiExperiments) [inline interface/src/avatar/OtherAvatar.h:102]: this is a bit of a hack, maybe there's a better way to tell "this OtherAvatar doesn't have any avatar entities" (as opposed to "we just haven't received them from the avatar mixer yet")

File: libraries/networking/src/ResourceCache.h [hunk 4]
```diff
@@ -537,7 +531,7 @@ protected slots:
     bool _failedToLoad = false;
     bool _loaded = false;
 
-    QHash<QPointer<QObject>, float> _loadPriorities;
+    QHash<QPointer<QObject>, std::function<float()>> _loadPriorityOperators;
```
Comment (HifiExperiments) [inline libraries/networking/src/ResourceCache.h:534]: before, each resource had a set priority, which never updated.  so if you went to a heavy domain, and moved around while things were loading, the priorities wouldn't change, so the loading would happen based on your original position.  now, we save these operators instead, so we can reevaluate the priority as we go

**Comment by JulianGro** [comment]
I assume `waitForWearables` is global in the sense that is affects all wearables an avatar has? I feel like we should be able to toggle this per wearable. I might want to wait for basic clothing to load, but a hat or some other accessory is relatively unimportant and can take its time.

**Comment by HifiExperiments** [comment]
@JulianGro that's correct, `waitForWearables` waits for all wearables.  hm...making it per-entity might be a bit complicated because clients don't really know how many avatar entities to expect until they've actually arrived.  do you envision a property that you set on the entities themselves or some way to specify which ones to wait for in the FST?  I also worry that making it too complicated will deter people from using it, seems easier to just have a single flag that needs to be set.

**Comment by JulianGro** [comment]
I don't really know how you specify wearables in the FST. I assumed it is just a list of other files to load; In that list, you could add that property for every entry of the list.

**Comment by HifiExperiments** [comment]
oh!  I didn’t know you could do that.  I think I see it in the code now.  yeah I think I can make it so FST-defined entities can add a flag to just those that you actually want to wait for.  maybe we also keep the global waitForWearables as a second option, especially since not all entities are defined like that?

**Comment by ksuprynowicz** [comment]
> maybe we also keep the global waitForWearables as a second option, especially since not all entities are defined like that?

I think it's a really good idea.

**Comment by HifiExperiments** [comment]
> I think I see it in the code now.

I was wrong lol; we support loading avatar entities from bookmarks, but not FSTs directly.  that would definitely be a nice feature to have, but maybe we handle it in separate PR (along with being able to specify those individual ones to wait for)?

**Review by ksuprynowicz — APPROVED** [review]
Everything looks good :)

**Comment by ksuprynowicz** [comment]
I just tested it on Linux and something very weird is happening audio and avatar mixers are crashing:
```
[08/18 16:48:10] [DEBUG] [hifi.networking] [62380] [audio-mixer] unreasonable sequence number: 0 (possible duplicate)
[08/18 16:48:10] [DEBUG] [hifi.networking] [62380] [audio-mixer] unreasonable sequence number: 1 (possible duplicate)
[08/18 16:48:10] [DEBUG] [hifi.networking] [62380] [audio-mixer] unreasonable sequence number: 2 (possible duplicate)
[08/18 16:48:10] [DEBUG] [hifi.networking] [62380] [audio-mixer] unreasonable sequence number: 3 (possible duplicate)
[08/18 16:48:10] [DEBUG] [hifi.networking] [62380] [audio-mixer] unreasonable sequence number: 4 (possible duplicate)
[08/18 16:48:10] [DEBUG] [hifi.networking] [62380] [audio-mixer] unreasonable sequence number: 5 (possible duplicate)
[08/18 16:48:10] [DEBUG] [hifi.networking] [62380] [audio-mixer] unreasonable sequence number: 6 (possible duplicate)
[08/18 16:48:10] [DEBUG] [hifi.networking] [62380] [audio-mixer] unreasonable sequence number: 7 (possible duplicate)
[08/18 16:48:10] [DEBUG] [hifi.networking] [62380] [audio-mixer] re-synced with sequence number sender
assignment-client: /home/ksuprynowicz/overte/overte_v8/overte/libraries/entities/src/EntityTree.cpp:2074: int EntityTree::processEraseMessage(ReceivedMessage&, const SharedNodePointer&): Assertion `!getIsServer()' failed.
Aborted (core dumped)
```
and
```
[08/18 16:48:09] [DEBUG] [hifi.networking] [62388] [avatar-mixer] Added "Agent" (I) {53ed13e6-102c-4475-b7c9-4462743e39c1}(30294) "UDP ""89.64.89.53":61311 / "UDP ""192.168.0.15":58793 
[08/18 16:48:09] [DEBUG] [hifi.networking] [62388] [avatar-mixer] Creating new Connection class for "UDP ""192.168.0.15":58793
[08/18 16:48:09] [DEBUG] [hifi.networking] [62388] [avatar-mixer] Creating new Connection class for "UDP ""89.64.89.53":61311
[08/18 16:48:09] [DEBUG] [hifi.networking] [62388] [avatar-mixer] Packet of type 3 (Ping) received from unknown node with Local ID 18111 
[08/18 16:48:09] [DEBUG] [hifi.networking] [62388] [avatar-mixer] Activating local socket for network peer with ID "53ed13e6-102c-4475-b7c9-4462743e39c1"
[08/18 16:48:10] [DEBUG] [hifi.avatars] [62388] [avatar-mixer] Processing first identity packet for QUuid("{53ed13e6-102c-4475-b7c9-4462743e39c1}") - 2
[08/18 16:48:10] [DEBUG] [hifi.avatars] [62388] [avatar-mixer] Giving session display name "74hc595" to node with ID QUuid("{53ed13e6-102c-4475-b7c9-4462743e39c1}")
[08/18 16:48:10] [SUPPRESS] [] [62388] [avatar-mixer] 3 repeated log entries - Last entry: "Packet of type 3 (Ping) received from unknown node with Local ID 18111 "
assignment-client: /home/ksuprynowicz/overte/overte_v8/overte/libraries/entities/src/EntityTree.cpp:2074: int EntityTree::processEraseMessage(ReceivedMessage&, const SharedNodePointer&): Assertion `!getIsServer()' failed.
Aborted (core dumped)
```

**Comment by ksuprynowicz** [comment]
I'm not sure if it's a bug , but it shows NaN priority for loading wearables:
overte-snap-by-X74hc595-on-2024-08-18_17-41-00

**Comment by ksuprynowicz** [comment]
Next time I tried loading it priority was not NaN anymore.
It looks like waitForWearables is not working though. The avatar (https://oaktown.pl/hazraptor_1/avatar_shark_3a_wearables.fst) already shows up while wearable(https://data.moto9000.moe/domain/Tile-59-68-1-1.obj) is still loading.
overte-snap-by-X74hc595-on-2024-08-18_17-47-02

**Comment by HifiExperiments** [comment]
thank you @ksuprynowicz!  that huge model made it a lot easier to test this, `waitForWearables` should be working now!

the assert was actually due to a change I made in the [sound entities PR](https://github.com/overte-org/overte/pull/894/files#diff-641be998f47662089055d6bdf9bed4edde258f100422ea36cab336ec82f5b4e8R819), I started marking more of the mixers properly as "servers", but some of the logic actually needed to be specifically for "isEntityServer".  I believe I've fixed that

the NaNs were somewhat expected.  if we were below the concurrent download limit, we just start downloading right away, and the asset doesn't get a priority at all.  but I've made it show it just doesn't show anything, rather than showing NaN, because that was confusing.

**Comment by ksuprynowicz** [comment]
I just tested it again, and everything works perfectly now :)
I think it's ready to merge.

**Pull request merged**